src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1006,251 +1083,422 @@ RouterSelectTask(Query *originalQuery, Query *query)
 }
 
 
 /*
- * ColumnMatchExpressionAtTopLevelConjunction returns true if the query contains an exact
- * match (equal) expression on the provided column. The function returns true only
- * if the match expression has an AND relation with the rest of the expression tree.
+ * RouterQueryJob creates a Job for the specified query to execute the
+ * provided single shard select task.
  */
-static bool
-ColumnMatchExpressionAtTopLevelConjunction(Node *node, Var *column)
+static Job *
+RouterQueryJob(Query *query, Task *task, List *placementList)
 {
-	if (node == NULL)
+	Job *job = NULL;
+	List *taskList = NIL;
+	TaskType taskType = task->taskType;
+
+	/*
+	 * We send modify task to the first replica, otherwise we choose the target shard
+	 * according to task assignment policy. Placement list for select queries are
+	 * provided as function parameter.
+	 */
+	if (taskType == MODIFY_TASK)
 	{
-		return false;
+		taskList = FirstReplicaAssignTaskList(list_make1(task));
 	}
-
-	if (IsA(node, OpExpr))
+	else
 	{
-		OpExpr *opExpr = (OpExpr *) node;
-		bool simpleExpression = SimpleOpExpression((Expr *) opExpr);
-		bool columnInExpr = false;
-		bool usingEqualityOperator = false;
+		Assert(placementList != NIL);
 
-		if (!simpleExpression)
-		{
-			return false;
-		}
+		task->taskPlacementList = placementList;
+		taskList = list_make1(task);
+	}
 
-		columnInExpr = OpExpressionContainsColumn(opExpr, column);
-		if (!columnInExpr)
-		{
-			return false;
-		}
+	job = CitusMakeNode(Job);
+	job->dependedJobList = NIL;
+	job->jobId = INVALID_JOB_ID;
+	job->subqueryPushdown = false;
+	job->jobQuery = query;
+	job->taskList = taskList;
 
-		usingEqualityOperator = OperatorImplementsEquality(opExpr->opno);
+	return job;
+}
 
-		return usingEqualityOperator;
+
+/*
+ * MultiRouterPlannableQuery returns true if given query can be router plannable.
+ * The query is router plannable if it is a select query issued on a hash
+ * partitioned distributed table, and it has a exact match comparison on the
+ * partition column. This feature is enabled if task executor is set to real-time
+ */
+bool
+MultiRouterPlannableQuery(Query *query, MultiExecutorType taskExecutorType,
+						  RelationRestrictionContext *restrictionContext)
+{
+	CmdType commandType = query->commandType;
+	ListCell *relationRestrictionContextCell = NULL;
+
+	if (commandType == CMD_INSERT || commandType == CMD_UPDATE ||
+		commandType == CMD_DELETE)
+	{
+		return true;
 	}
-	else if (IsA(node, BoolExpr))
+
+	/* FIXME: I tend to think it's time to remove this */
+	if (taskExecutorType != MULTI_EXECUTOR_REAL_TIME)
 	{
-		BoolExpr *boolExpr = (BoolExpr *) node;
-		List *argumentList = boolExpr->args;
-		ListCell *argumentCell = NULL;
+		return false;
+	}
 
-		if (boolExpr->boolop != AND_EXPR)
-		{
-			return false;
-		}
+	Assert(commandType == CMD_SELECT);
+
+	/*
+	 * FIXME, figure out which restrictions we exactly want. FOR UPDATE seems
+	 * to make no sense, given the transactional behaviour.
```

Comment:
more like being removed ? I think we removed almost all of restriction in the logical planner.